gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/ResumeDagProc.java

```diff
@@ -0,0 +1,138 @@
+
+import static org.apache.gobblin.service.ExecutionStatus.CANCELLED;
+import static org.apache.gobblin.service.ExecutionStatus.FAILED;
+import static org.apache.gobblin.service.ExecutionStatus.PENDING_RESUME;
+
+
+/**
+ * An implementation of {@link DagProc} for resuming {@link DagTask}.
+ */
+
+@Alpha
+@Slf4j
+public final class ResumeDagProc extends DagProc {
+  private DagManager.DagId resumeDagId;
+  private DagManagementStateStore dagManagementStateStore;
+  private MetricContext metricContext;
+  private Optional<EventSubmitter> eventSubmitter;
+  private DagStateStore dagStateStore;
+  private DagStateStore failedDagStateStore;
+  private MultiActiveLeaseArbiter multiActiveLeaseArbiter;
+  private UserQuotaManager quotaManager;
+  private DagManagerMetrics dagManagerMetrics;
+
+  public ResumeDagProc(DagManager.DagId resumeDagId) throws InstantiationException, IllegalAccessException, IOException {
+    this.resumeDagId = resumeDagId;
+    //TODO: add this to dagproc factory instead
+    this.dagManagementStateStore = new DagManagementStateStore();
+    this.metricContext = Instrumented.getMetricContext(ConfigUtils.configToState(ConfigFactory.empty()), getClass());
+    this.multiActiveLeaseArbiter = new MysqlMultiActiveLeaseArbiter(ConfigBuilder.create().build());
+    this.eventSubmitter = Optional.of(new EventSubmitter.Builder(this.metricContext, "org.apache.gobblin.service").build());
+    this.dagStateStore = (DagStateStore) (MysqlDagStateStoreFactory.class.newInstance()).createStateStore(ConfigBuilder.create()
+        .build(), State.class);
+    this.failedDagStateStore = (DagStateStore) (MysqlDagStateStoreFactory.class.newInstance()).createStateStore(ConfigBuilder.create()
+        .build(), State.class);
+    this.quotaManager = GobblinConstructorUtils.invokeConstructor(UserQuotaManager.class,
+        ConfigUtils.getString(ConfigBuilder.create().build(), ServiceConfigKeys.QUOTA_MANAGER_CLASS, ServiceConfigKeys.DEFAULT_QUOTA_MANAGER),
+        ConfigBuilder.create().build());
+    this.dagManagerMetrics = new DagManagerMetrics();
+  }
+
+  @Override
+  protected Object initialize() throws IOException {
+    String dagIdToResume= resumeDagId.toString();
+    Dag<JobExecutionPlan> dag = this.failedDagStateStore.getDag(dagIdToResume);
+    return dag;
+  }
+
+  @Override
+  protected Object act(Object state) throws IOException {
+    String dagIdToResume= resumeDagId.toString();
+    Dag<JobExecutionPlan> dag = (Dag<JobExecutionPlan>) state;
+    if (!this.dagManagementStateStore.getFailedDagIds().contains(dagIdToResume)) {
+      log.warn("No dag found with dagId " + dagIdToResume + ", so cannot resume flow");
+      this.dagManagementStateStore.removeDagActionFromStore(resumeDagId, DagActionStore.FlowActionType.RESUME);
+      return dag;
+    }
+
+    if (dag == null) {
+      log.error("Dag " + dagIdToResume + " was found in memory but not found in failed dag state store");
```

Comment:
Wouldn't this be in the above check rather than here for this scenario? If Dag is null then whatever object was sent is invalid